packages/dev/core/src/Materials/materialHelper.functions.ts

```diff
@@ -598,6 +600,98 @@ export function GetFogState(mesh: AbstractMesh, scene: Scene) {
     return scene.fogEnabled && mesh.applyFog && scene.fogMode !== Constants.FOGMODE_NONE;
 }
 
+/**
+ * Interface representing metadata for vertex pulling
+ */
+export interface IVertexPullingMetadata {
+    /**
+     * Offset in vertex buffer where data starts
+     */
+    offset: number;
+
+    /**
+     * Stride between elements in the vertex buffer
+     */
+    stride: number;
+
+    /**
+     * Type of the vertex buffer (e.g., float, int)
+     */
+    type: number; // VertexBuffer type constant
+}
+
+// Store vertex pulling metadata per geometry
+const _VertexPullingMetadataCache = new WeakMap<Geometry, Map<string, IVertexPullingMetadata>>();
+
+/**
+ * Prepares vertex pulling uniforms for the given attributes and mesh
+ * @param geometry The geometry containing the vertex buffers
+ * @returns A map of attribute names to their metadata, or null if unavailable
+ */
+export function PrepareVertexPullingUniforms(geometry: Geometry): Nullable<Map<string, IVertexPullingMetadata>> {
+    if (!geometry) {
+        return null;
+    }
+    const vertexBuffers = geometry.getVertexBuffers();
+    if (!vertexBuffers) {
+        return null;
+    }
+
+    // Check cache first
+    let metadata = _VertexPullingMetadataCache.get(geometry);
+    if (!metadata) {
+        metadata = new Map<string, IVertexPullingMetadata>();
+        _VertexPullingMetadataCache.set(geometry, metadata);
+    } else {
+        // Return cached metadata if it exists and hasn't changed
+        let needsUpdate = false;
+        for (const vb in vertexBuffers) {
+            if (!metadata.has(vb)) {
+                needsUpdate = true;
+                break;
+            }
+        }
+        if (!needsUpdate) {
+            return metadata;
+        }
+    }
+
+    // Build or update metadata
+    for (const vb in vertexBuffers) {
+        const vertexBuffer = vertexBuffers[vb];
+        if (vertexBuffer) {
+            const offset = vertexBuffer.byteOffset;
+            const stride = vertexBuffer.byteStride;
+            const type = vertexBuffer.type;
+
+            metadata.set(vb, {
+                offset: offset,
+                stride: stride,
+                type: type,
+            });
+        }
+    }
+
+    return metadata;
+}
+
+/**
+ * Bind vertex pulling uniforms to the effect
+ * @param effect The effect to bind the uniforms to
+ * @param metadata The vertex pulling metadata
+ */
+export function BindVertexPullingUniforms(effect: Effect, metadata: Map<string, IVertexPullingMetadata>): void {
+    if (!metadata || !effect) {

```

Comment:
If **metadata** and **effect** can be null/undefined, you should update their declaration in `BindVertexPullingUniforms` accordingly.

Reply:
The metadata can't be null here. I've removed the check now.